specs/latest/2.0/index.html

```diff
@@ -3192,6 +3192,37 @@ <h2 class="no-toc">Table of contents</h2>
         <code>FLOAT_32_UNSIGNED_INT_24_8_REV</code>, generates an <code>INVALID_ENUM</code>.
     </p>
 
+    <h3>VertexAttrib function must match shader attribute type</h3>
+
+    <p>
+        If <code>vertexAttrib{1234}f</code> or <code>vertexAttrib{1234}fv</code> is used and the base
+        type of the shader at slot <code>index</code> is not floating-point (e.g. is signed or unsigned
+        integer), generates an <code>INVALID_OPERATION</code> error.
+    </p>
+    <p>
+        If <code>vertexAttribI4ui</code> or <code>vertexAttribI4uiv</code> is used and the base type of
+        the shader attribute at slot <code>index</code> is floating-point or signed integer,
+        generates an <code>INVALID_OPERATION</code> error.
+    </p>
+    <p>
+        If <code>vertexAttribI4i</code> or <code>vertexAttribI4iv</code> is used and the base type of
+        the shader attribute at slot <code>index</code> is floating-point or unsigned integer,
+        generates an <code>INVALID_OPERATION</code> error.
+    </p>
```

Comment:
I don't see how it will be possible to enforce these; the vertex attribute state is part of that of the bound vertex array object, not the program object. Therefore, calls to UseProgram or BindVertexArray will change this state, so an error can't be generated upon call to vertexAttrib*. Am I missing something?